sdk/storage/AzureStorageBlob/Source/DataStructures/StorageStreamDownloader.swift

```diff
@@ -318,22 +329,41 @@ public class BlobStreamDownloader {
     ///   - completion: A completion handler called when the download completes.
     public func complete(inGroup group: DispatchGroup? = nil, then completion: @escaping () -> Void) throws {
         guard !isComplete else {
-            completion()
+            let progress = BlobDownloadProgress(bytes: self.progress, totalBytes: self.fileSize ?? 1)
+            if let delegate = self.delegate {
+                delegate.downloader(self, didFinishWithProgress: progress)
+            } else {
+                completion()
+            }
             return
         }
+        let defaultGroup = DispatchGroup()
+        let dispatchGroup = group ?? defaultGroup
         for _ in blockList {
-            group?.enter()
-            next(inGroup: group)
+            dispatchGroup.enter()
+            next(inGroup: dispatchGroup) { result, response in
+                if let delegate = self.delegate {
+                    let test = "\(result) \(response)"
+                    let progress = BlobDownloadProgress(bytes: self.progress, totalBytes: self.fileSize ?? 1)
+                    delegate.downloader(self, didUpdateWithProgress: progress)
+                }
+            }
         }
-        group?.notify(queue: DispatchQueue.main) {
-            completion()
+        dispatchGroup.notify(queue: DispatchQueue.main) {
+            let progress = BlobDownloadProgress(bytes: self.progress, totalBytes: self.fileSize ?? 1)
+            if let delegate = self.delegate {
+                delegate.downloader(self, didFinishWithProgress: progress)
+            } else {
+                completion()
+            }
         }
     }
 
     /// Download the contents of this file to a stream.
     /// - Parameters:
     ///   - group: An optional `DispatchGroup` to wait for the download to complete.
-    public func next(inGroup group: DispatchGroup? = nil) {
+    ///   - completion: A completion handler with which to process the downloaded chunk.
+    public func next(inGroup group: DispatchGroup? = nil, then completion: @escaping (Result<Data, Error>, HTTPResponse) -> Void) {
```

Comment:
The type for `completion` should be `HTTPResultHandler<Data>` (the type alias just makes things simpler).

Reply:
Fixed.